plugins/inputs/aliyuncms/discovery.go

```diff
@@ -0,0 +1,508 @@
+	}
+
+	return
+}
+
+func (dt *discoveryTool) getDiscoveryData(cli aliyunSdkClient, req *requests.CommonRequest, limiter chan bool) (map[string]interface{}, error) {
+	var (
+		err           error
+		resp          *responses.CommonResponse
+		data          []interface{}
+		discoveryData []interface{}
+		totalCount    int
+		pageNumber    int
+	)
+	defer delete(req.QueryParams, "PageNumber")
+
+	for {
+		if limiter != nil {
+			<-limiter //Rate limiting
+		}
+
+		resp, err = cli.ProcessCommonRequest(req)
+		if err != nil {
+			return nil, err
+		}
+
+		data, totalCount, _, pageNumber, err = dt.parseDiscoveryResponse(resp)
+		if err != nil {
+			return nil, err
+		}
+		discoveryData = append(discoveryData, data...)
+
+		//Pagination
+		pageNumber++
+		req.QueryParams["PageNumber"] = strconv.Itoa(pageNumber)
+
+		if len(discoveryData) == totalCount { //All data received
+			//Map data to appropriate shape before return
+			preparedData := map[string]interface{}{}
+
+			for _, raw := range discoveryData {
+				if elem, ok := raw.(map[string]interface{}); ok {
+					if objectId, ok := elem[dt.respObjectIdKey].(string); ok {
+						preparedData[objectId] = elem
+					}
+				} else {
+					return nil, errors.Errorf("Can't parse input data element, not a map[string]interface{} type")
+				}
+
+			}
+
+			return preparedData, nil
+		}
+
+	}
+
+}
+
+func (dt *discoveryTool) getDiscoveryDataAllRegions(limiter chan bool) (map[string]interface{}, error) {
+	var (
+		data       map[string]interface{}
+		resultData = map[string]interface{}{}
+	)
+
+	for region, cli := range dt.cli {
+		//Building common request, as the code below is the same no matter
+		//which aliyun object type (project) is used
+		dscReq, ok := dt.req[region]
+		if !ok {
+			return nil, errors.Errorf("Error building common discovery request: not valid region %q", region)
+		}
+
+		rpcReq, err := getRpcReqFromDiscoveryRequest(dscReq)
+		if err != nil {
+			return nil, err
+		}
+
+		commonRequest := requests.NewCommonRequest()
+		commonRequest.Method = rpcReq.GetMethod()
+		commonRequest.Product = rpcReq.GetProduct()
+		commonRequest.Domain = rpcReq.GetDomain()
+		commonRequest.Version = rpcReq.GetVersion()
+		commonRequest.Scheme = rpcReq.GetScheme()
+		commonRequest.ApiName = rpcReq.GetActionName()
+		commonRequest.QueryParams = rpcReq.QueryParams
+		commonRequest.QueryParams["PageSize"] = strconv.Itoa(dt.reqDefaultPageSize)
+		commonRequest.TransToAcsRequest()
+
+		//Get discovery data using common request
+		data, err = dt.getDiscoveryData(cli, commonRequest, limiter)
+		if err != nil {
+			return nil, err
+		}
+
+		for k, v := range data {
+			resultData[k] = v
+		}
+	}
+	return resultData, nil
+}
+
+func (dt *discoveryTool) Start() {
+	var (
+		err      error
+		data     map[string]interface{}
+		lastData map[string]interface{}
+	)
+
+	//Initializing channel
+	dt.done = make(chan bool)
+
+	dt.wg.Add(1)
+	go func() {
+		defer dt.wg.Done()
+
+		ticker := time.NewTicker(dt.interval)
```

Comment:
Why start another internal service to do the discovery when you could potentially do it in the Gather function?

Reply:
Well, discovery process is independent from plugin `gather` as they gathers quite a different info (metrics vs discovery data). Discovery runs on it's own pace, as I commented earlier about the reasons why have `interval` and `discovery_interval`